File: backend/app/services/warehouse_service.py
"""
Warehouse Service — dashboard KPIs, activity feed, pending tasks, inventory grid.

Aggregates data from stock, movements, audits, and staging for the
warehouse dashboard and inventory views.
"""

from __future__ import annotations

import logging
from typing import Any

import aiosqlite

from app.models.warehouse import (
    ActivitySummary,
    DashboardData,
    DashboardKPIs,
    PendingTask,
    StagingGroup,
    StagingItem,
    WarehouseInventoryItem,
)

logger = logging.getLogger(__name__)


class WarehouseService:
    """High-level warehouse queries for dashboard and inventory pages."""

    def __init__(self, db: aiosqlite.Connection) -> None:
        self.db = db

    # ── Dashboard ─────────────────────────────────────────────────

    async def get_dashboard(self, show_dollars: bool = False) -> DashboardData:
        """Combined dashboard payload in a single call."""
        kpis = await self.get_kpis(show_dollars=show_dollars)
        activity = await self.get_recent_activity(limit=10)
        tasks = await self.get_pending_tasks()
        return DashboardData(kpis=kpis, recent_activity=activity, pending_tasks=tasks)

    async def get_kpis(self, show_dollars: bool = False) -> DashboardKPIs:
        """Calculate dashboard KPI numbers."""
        # Stock health %: parts that have warehouse qty within min-max range
        cursor = await self.db.execute(
            """SELECT
                   COUNT(DISTINCT p.id) AS total_parts,
                   COUNT(DISTINCT CASE
                       WHEN COALESCE(s.wh_qty, 0) >= p.min_stock_level
                        AND (p.max_stock_level = 0 OR COALESCE(s.wh_qty, 0) <= p.max_stock_level)
                       THEN p.id
                   END) AS healthy_parts
               FROM parts p
               LEFT JOIN (
                   SELECT part_id, SUM(qty) AS wh_qty
                   FROM stock WHERE location_type = 'warehouse'
                   GROUP BY part_id
               ) s ON s.part_id = p.id
               WHERE p.is_deprecated = 0
                 AND p.target_stock_level > 0"""
        )
        row = await cursor.fetchone()
        total_parts = row["total_parts"] or 0
        healthy_parts = row["healthy_parts"] or 0
        health_pct = round((healthy_parts / total_parts * 100) if total_parts > 0 else 100, 1)

        # Total units in warehouse
        cursor = await self.db.execute(
            "SELECT COALESCE(SUM(qty), 0) AS total FROM stock WHERE location_type = 'warehouse'"
        )
        row = await cursor.fetchone()
        total_units = row["total"] if row else 0

        # Warehouse value (only if allowed)
        warehouse_value = None
        if show_dollars:
            cursor = await self.db.execute(
                """SELECT COALESCE(SUM(s.qty * p.company_cost_price), 0) AS val
                   FROM stock s JOIN parts p ON p.id = s.part_id
                   WHERE s.location_type = 'warehouse'"""
            )
            row = await cursor.fetchone()
            warehouse_value = round(row["val"], 2) if row else 0.0

        # Shortfall count (parts below min, excluding winding-down)
        cursor = await self.db.execute(
            """SELECT COUNT(DISTINCT p.id) AS cnt
               FROM parts p
               LEFT JOIN (
                   SELECT part_id, SUM(qty) AS wh_qty
                   FROM stock WHERE location_type = 'warehouse'
                   GROUP BY part_id
               ) s ON s.part_id = p.id
               WHERE p.is_deprecated = 0
                 AND p.target_stock_level > 0
                 AND p.min_stock_level > 0
                 AND COALESCE(s.wh_qty, 0) < p.min_stock_level"""
        )
        row = await cursor.fetchone()
        shortfall = row["cnt"] if row else 0

        # Pending tasks count
        tasks = await self.get_pending_tasks()

        return DashboardKPIs(
            stock_health_pct=health_pct,
            total_units=total_units,
            warehouse_value=warehouse_value,
            shortfall_count=shortfall,
            pending_task_count=len(tasks),
        )

    async def get_recent_activity(self, limit: int = 10) -> list[ActivitySummary]:
        """Get recent movements as one-line summaries."""
        cursor = await self.db.execute(
            """SELECT m.id, m.movement_type, m.qty,
                      m.from_location_type, m.to_location_type,
                      m.created_at,
                      p.name AS part_name,
                      u.display_name AS performer_name
               FROM stock_movements m
               JOIN parts p ON p.id = m.part_id
               LEFT JOIN users u ON u.id = m.performed_by
               ORDER BY m.created_at DESC
               LIMIT ?""",
            (limit,),
        )
        rows = await cursor.fetchall()

        results = []
        for row in rows:
            # Build one-line summary
            action = _movement_verb(row["movement_type"])
            user = row["performer_name"] or "Someone"
            summary = f"{user} {action} {row['qty']}× {row['part_name']}"

            if row["from_location_type"] and row["to_location_type"]:
                summary += f" ({row['from_location_type']} → {row['to_location_type']})"

            results.append(ActivitySummary(
                id=row["id"],
                summary=summary,
                movement_type=row["movement_type"],
                performer_name=row["performer_name"],
                created_at=row["created_at"],
            ))

        return results

    async def get_pending_tasks(self) -> list[PendingTask]:
        """Get all pending warehouse tasks."""
        tasks: list[PendingTask] = []

        # 1. Staged items awaiting pickup (grouped by destination)
        cursor = await self.db.execute(
            """SELECT st.destination_type, st.destination_id, st.destination_label,
                      COUNT(*) AS item_count,
                      SUM(s.qty) AS total_qty,
                      MIN(st.created_at) AS oldest,
                      ROUND((julianday('now') - julianday(MIN(st.created_at))) * 24, 1) AS hours_old
               FROM staging_tags st
               JOIN stock s ON s.id = st.stock_id AND s.qty > 0
               GROUP BY st.destination_type, st.destination_id"""
        )
        for row in await cursor.fetchall():
            hours = row["hours_old"] or 0
            severity = "critical" if hours > 48 else "warning" if hours > 24 else "normal"
            label = row["destination_label"] or f"{row['destination_type']} #{row['destination_id']}"

            tasks.append(PendingTask(
                task_type="staged_item",
                title=f"Staged: {row['total_qty']} units for {label}",
                subtitle=f"{row['item_count']} items, {_format_hours(hours)}",
                severity=severity,
                destination_type=row["destination_type"],
                destination_id=row["destination_id"],
            ))

        # 2. Active/paused audits
        cursor = await self.db.execute(
            """SELECT a.id, a.audit_type, a.status,
                      a.total_items, a.matched_items, a.discrepancy_count,
                      pc.name AS category_name
               FROM audits a
               LEFT JOIN part_categories pc ON pc.id = a.category_id
               WHERE a.status IN ('in_progress', 'paused')
               ORDER BY a.created_at DESC"""
        )
        for row in await cursor.fetchall():
            counted = (row["matched_items"] or 0) + (row["discrepancy_count"] or 0)
            total = row["total_items"] or 0
            pct = round(counted / total * 100) if total > 0 else 0

            type_label = row["audit_type"].replace("_", " ").title()
            if row["category_name"]:
                type_label += f": {row['category_name']}"

            tasks.append(PendingTask(
                task_type="audit",
                title=f"{type_label} — {pct}%",
                subtitle=f"{counted}/{total} items counted" if total > 0 else "Starting...",
                severity="normal",
                audit_id=row["id"],
            ))

        # 3. Spot-check requests (pending low-stock verification)
        cursor = await self.db.execute(
            """SELECT a.id AS audit_id, ai.part_id, p.name AS part_name,
                      ai.expected_qty
               FROM audits a
               JOIN audit_items ai ON ai.audit_id = a.id
               JOIN parts p ON p.id = ai.part_id
               WHERE a.audit_type = 'spot_check'
                 AND a.status = 'in_progress'
                 AND ai.result = 'pending'
               ORDER BY a.created_at DESC"""
        )
        for row in await cursor.fetchall():
            tasks.append(PendingTask(
                task_type="spot_check",
                title=f"Verify Count: {row['part_name']}",
                subtitle=f"System shows {row['expected_qty']} units",
                severity="warning",
                audit_id=row["audit_id"],
                part_id=row["part_id"],
            ))

        return tasks

    # ── Inventory Grid ────────────────────────────────────────────

    async def get_warehouse_inventory(
        self,
        *,
        search: str | None = None,
        category_id: int | None = None,
        brand_id: int | None = None,
        part_id: int | None = None,
        stock_status: str | None = None,
        sort_by: str = "name",
        sort_dir: str = "asc",
        page: int = 1,
        page_size: int = 50,
        show_dollars: bool = False,
    ) -> tuple[list[WarehouseInventoryItem], int]:
        """Paginated warehouse inventory with health calculations.

        Only shows "warehouse-relevant" parts: those with actual stock
        or a non-zero target level. Parts with 0 stock AND 0 target
        (wound-out / never stocked) are excluded.
        """
        where_clauses = [
            "p.is_deprecated = 0",
            # Only warehouse-relevant: has stock or has a target
            "(COALESCE(wh.qty, 0) > 0 OR p.target_stock_level > 0)",
        ]
        params: list[Any] = []

        if part_id:
            where_clauses.append("p.id = ?")
            params.append(part_id)
        if search:
            where_clauses.append(
                "(p.name LIKE ? OR p.code LIKE ? OR p.shelf_location LIKE ? OR p.bin_location LIKE ?)"
            )
            params.extend([f"%{search}%"] * 4)
        if category_id:
            where_clauses.append("p.category_id = ?")
            params.append(category_id)
        if brand_id:
            where_clauses.append("p.brand_id = ?")
            params.append(brand_id)

        # Stock status filter — use WHERE on the subquery expression
        # (HAVING requires GROUP BY; these are non-aggregate filters on a computed column)
        if stock_status == "low_stock":
            where_clauses.append("COALESCE(wh.qty, 0) < p.min_stock_level AND p.target_stock_level > 0")
        elif stock_status == "overstock":
            where_clauses.append("p.max_stock_level > 0 AND COALESCE(wh.qty, 0) > p.max_stock_level")
        elif stock_status == "winding_down":
            # Has stock we don't want: target is 0 but warehouse qty > 0
            where_clauses.append("p.target_stock_level = 0 AND COALESCE(wh.qty, 0) > 0")
        elif stock_status == "zero":
            where_clauses.append("COALESCE(wh.qty, 0) = 0 AND p.target_stock_level > 0")
        elif stock_status == "in_range":
            where_clauses.append(
                "COALESCE(wh.qty, 0) >= p.min_stock_level "
                "AND (p.max_stock_level = 0 OR COALESCE(wh.qty, 0) <= p.max_stock_level) "
                "AND p.target_stock_level > 0"
            )

        where_sql = " AND ".join(where_clauses)

        # Sort mapping
        sort_map = {
            "name": "p.name",
            "code": "p.code",
            "category": "pc.name",
            "brand": "b.name",
            "qty": "wh_qty",
            "shelf": "p.shelf_location",
        }
        order_col = sort_map.get(sort_by, "p.name")
        order_dir = "DESC" if sort_dir.lower() == "desc" else "ASC"

        # Count query
        count_sql = f"""
            SELECT COUNT(*) AS cnt
            FROM parts p
            LEFT JOIN (
                SELECT part_id, SUM(qty) AS qty
                FROM stock WHERE location_type = 'warehouse'
                GROUP BY part_id
            ) wh ON wh.part_id = p.id
            WHERE {where_sql}
        """
        cursor = await self.db.execute(count_sql, params)
        row = await cursor.fetchone()
        total = row["cnt"] if row else 0

        # Data query
        offset = (page - 1) * page_size
        data_sql = f"""
            SELECT
                p.id AS part_id,
                p.code AS part_code,
                p.name AS part_name,
                p.category_id,
                pc.name AS category_name,
                p.brand_id,
                b.name AS brand_name,
                p.unit_of_measure,
                p.shelf_location,
                p.bin_location,
                COALESCE(wh.qty, 0) AS warehouse_qty,
                COALESCE(pulled.qty, 0) AS pulled_qty,
                COALESCE(truck.qty, 0) AS truck_qty,
                COALESCE(wh.qty, 0) + COALESCE(pulled.qty, 0) + COALESCE(truck.qty, 0) AS total_qty,
                p.min_stock_level,
                p.target_stock_level,
                p.max_stock_level,
                p.company_cost_price,
                p.forecast_days_until_low,
                p.is_qr_tagged,
                COALESCE(wh.qty, 0) AS wh_qty
            FROM parts p
            LEFT JOIN part_categories pc ON pc.id = p.category_id
            LEFT JOIN brands b ON b.id = p.brand_id
            LEFT JOIN (
                SELECT part_id, SUM(qty) AS qty FROM stock
                WHERE location_type = 'warehouse' GROUP BY part_id
            ) wh ON wh.part_id = p.id
            LEFT JOIN (
                SELECT part_id, SUM(qty) AS qty FROM stock
                WHERE location_type = 'pulled' GROUP BY part_id
            ) pulled ON pulled.part_id = p.id
            LEFT JOIN (
                SELECT part_id, SUM(qty) AS qty FROM stock
                WHERE location_type = 'truck' GROUP BY part_id
            ) truck ON truck.part_id = p.id
            WHERE {where_sql}
            ORDER BY {order_col} {order_dir}
            LIMIT ? OFFSET ?
        """
        cursor = await self.db.execute(data_sql, (*params, page_size, offset))
        rows = await cursor.fetchall()

        items = []
        for row in rows:
            wh = row["warehouse_qty"]
            min_s = row["min_stock_level"] or 0
            max_s = row["max_stock_level"] or 0
            target = row["target_stock_level"] or 0

            # Determine stock status
            if target == 0:
                status = "winding_down"
            elif wh == 0:
                status = "zero"
            elif min_s > 0 and wh < min_s:
                status = "low_stock"
            elif max_s > 0 and wh > max_s:
                status = "overstock"
            else:
                status = "in_range"

            # Health percentage (for progress bar)
            if max_s > 0:
                health_pct = min(100, round(wh / max_s * 100, 1))
            elif target > 0:
                health_pct = min(100, round(wh / target * 100, 1))
            else:
                health_pct = 100.0

            items.append(WarehouseInventoryItem(
                part_id=row["part_id"],
                part_code=row["part_code"],
                part_name=row["part_name"],
                category_id=row["category_id"],
                category_name=row["category_name"],
                brand_id=row["brand_id"],
                brand_name=row["brand_name"],
                unit_of_measure=row["unit_of_measure"],
                shelf_location=row["shelf_location"],
                bin_location=row["bin_location"],
                warehouse_qty=wh,
                pulled_qty=row["pulled_qty"],
                truck_qty=row["truck_qty"],
                total_qty=row["total_qty"],
                min_stock_level=min_s,
                target_stock_level=target,
                max_stock_level=max_s,
                stock_status=status,
                health_pct=health_pct,
                unit_cost=row["company_cost_price"] if show_dollars else None,
                total_value=round(wh * (row["company_cost_price"] or 0), 2) if show_dollars else None,
                forecast_days_until_low=row["forecast_days_until_low"],
                is_qr_tagged=bool(row["is_qr_tagged"]),
            ))

        return items, total

    # ── Staging ───────────────────────────────────────────────────

    async def get_staging_groups(self) -> list[StagingGroup]:
        """Get all pulled/staged items grouped by destination."""
        cursor = await self.db.execute(
            """SELECT s.id AS stock_id, s.part_id, s.qty,
                      p.name AS part_name, p.code AS part_code,
                      sup.name AS supplier_name,
                      st.destination_type, st.destination_id, st.destination_label,
                      st.created_at AS staged_at,
                      u.display_name AS tagged_by_name,
                      ROUND((julianday('now') - julianday(st.created_at)) * 24, 1) AS hours_staged
               FROM stock s
               JOIN parts p ON p.id = s.part_id
               LEFT JOIN suppliers sup ON sup.id = s.supplier_id
               LEFT JOIN staging_tags st ON st.stock_id = s.id
               LEFT JOIN users u ON u.id = st.tagged_by
               WHERE s.location_type = 'pulled' AND s.qty > 0
               ORDER BY st.destination_type, st.destination_id, p.name"""
        )
        rows = await cursor.fetchall()

        # Group by destination
        groups: dict[str, StagingGroup] = {}
        for row in rows:
            dest_type = row["destination_type"]
            dest_id = row["destination_id"]

            if dest_type and dest_id:
                key = f"{dest_type}:{dest_id}"
                label = row["destination_label"] or f"{dest_type.title()} #{dest_id}"
            else:
                key = "untagged"
                label = "Untagged"

            if key not in groups:
                groups[key] = StagingGroup(
                    destination_type=dest_type,
                    destination_id=dest_id,
                    destination_label=label,
                )

            hours = row["hours_staged"] or 0
            aging = "critical" if hours > 48 else "warning" if hours > 24 else "normal"

            groups[key].items.append(StagingItem(
                stock_id=row["stock_id"],
                part_id=row["part_id"],
                part_name=row["part_name"],
                part_code=row["part_code"],
                qty=row["qty"],
                supplier_name=row["supplier_name"],
                destination_type=dest_type,
                destination_id=dest_id,
                destination_label=row["destination_label"],
                tagged_by_name=row["tagged_by_name"],
                staged_at=row["staged_at"],
                hours_staged=hours,
                aging_status=aging,
            ))
            groups[key].total_qty += row["qty"]
            if hours > groups[key].oldest_hours:
                groups[key].oldest_hours = hours
                groups[key].aging_status = aging

        return list(groups.values())

    # ── Parts Search (Context-Aware) ──────────────────────────────

    async def search_parts_for_wizard(
        self,
        query: str,
        location_type: str | None = None,
        location_id: int | None = None,
        limit: int = 20,
    ) -> list[dict]:
        """Search parts for the wizard, optionally scoped to a location.

        If location_type/id given, only returns parts WITH stock there.
        Otherwise returns all catalog parts.
        """
        if location_type and location_id is not None:
            # Context-aware: only parts with stock at source
            cursor = await self.db.execute(
                """SELECT p.id, p.code, p.name, p.image_url,
                          p.unit_of_measure, p.company_cost_price,
                          COALESCE(SUM(s.qty), 0) AS available_qty,
                          GROUP_CONCAT(DISTINCT sup.name) AS supplier_names
                   FROM parts p
                   JOIN stock s ON s.part_id = p.id
                       AND s.location_type = ? AND s.location_id = ? AND s.qty > 0
                   LEFT JOIN suppliers sup ON sup.id = s.supplier_id
                   WHERE (p.name LIKE ? OR p.code LIKE ?)
                     AND p.is_deprecated = 0
                   GROUP BY p.id
                   ORDER BY p.name
                   LIMIT ?""",
                (location_type, location_id, f"%{query}%", f"%{query}%", limit),
            )
        else:
            # All catalog parts
            cursor = await self.db.execute(
                """SELECT p.id, p.code, p.name, p.image_url,
                          p.unit_of_measure, p.company_cost_price,
                          COALESCE(wh.qty, 0) AS available_qty,
                          NULL AS supplier_names
                   FROM parts p
                   LEFT JOIN (
                       SELECT part_id, SUM(qty) AS qty FROM stock
                       WHERE location_type = 'warehouse' GROUP BY part_id
                   ) wh ON wh.part_id = p.id
                   WHERE (p.name LIKE ? OR p.code LIKE ?)
                     AND p.is_deprecated = 0
                   ORDER BY p.name
                   LIMIT ?""",
                (f"%{query}%", f"%{query}%", limit),
            )

        return await cursor.fetchall()


# ── Helpers ───────────────────────────────────────────────────────

def _movement_verb(movement_type: str) -> str:
    """Convert movement type to past-tense verb for activity feed."""
    return {
        "transfer": "moved",
        "consume": "consumed",
        "return": "returned",
        "receive": "received",
        "adjust": "adjusted",
        "write_off": "wrote off",
    }.get(movement_type, "moved")


def _format_hours(hours: float) -> str:
    """Format hours as a human-readable time string."""
    if hours < 1:
        return "just now"
    elif hours < 24:
        return f"{int(hours)}h ago"
    else:
        days = int(hours / 24)
        return f"{days}d ago"

File: frontend/src/local/services/warehouse-service.ts
/**
 * Local Warehouse Service — inventory read-only access for offline use.
 *
 * Mirrors the read portions of the warehouse/inventory services.
 * Field workers can browse inventory, check stock levels, and search
 * parts — all from the local database.
 *
 * Write operations (receiving, adjustments) are handled by movement-service.
 * Cost/value calculations are shop-only (field workers see quantities, not dollars).
 */

import { getDb } from '../db';

// ── Types ──────────────────────────────────────────────────────────

export interface DashboardKPIs {
  stock_health_pct: number;
  total_units: number;
  shortfall_count: number;
  pending_task_count: number;
}

export interface ActivitySummary {
  id: number;
  summary: string;
  movement_type: string;
  performer_name: string | null;
  created_at: string;
}

export interface InventoryItem {
  id: number;
  part_number: string;
  description: string;
  category_name: string | null;
  brand_name: string | null;
  warehouse_qty: number;
  min_stock: number;
  max_stock: number;
  reorder_point: number;
  status: string;
  last_moved: string | null;
}

export interface PartStockDetail {
  part_id: number;
  part_number: string;
  description: string;
  locations: StockLocation[];
  total_qty: number;
}

export interface StockLocation {
  location_type: string;
  location_id: number;
  location_name: string;
  qty: number;
  supplier_name: string | null;
}

// ── Service Functions ──────────────────────────────────────────────

/** Get warehouse dashboard KPIs */
export async function getDashboardKPIs(): Promise<DashboardKPIs> {
  const db = await getDb();

  // Total warehouse units
  const totalResult = await db.query(
    `SELECT COALESCE(SUM(qty), 0) as total FROM stock WHERE location_type = 'warehouse'`,
  );
  const totalUnits = totalResult.values[0]?.total ?? 0;

  // Stock health: parts within min-max range
  const healthResult = await db.query(
    `SELECT
       COUNT(*) as total_parts,
       SUM(CASE WHEN wh_qty >= p.min_stock_level AND wh_qty <= p.max_stock_level THEN 1 ELSE 0 END) as healthy
     FROM (
       SELECT part_id, COALESCE(SUM(qty), 0) as wh_qty
       FROM stock WHERE location_type = 'warehouse'
       GROUP BY part_id
     ) sq
     JOIN parts p ON p.id = sq.part_id
     WHERE p.is_active = 1`,
  );
  const totalParts = healthResult.values[0]?.total_parts ?? 1;
  const healthy = healthResult.values[0]?.healthy ?? 0;
  const healthPct = totalParts > 0 ? Math.round((healthy / totalParts) * 100) : 100;

  // Shortfall count (below min)
  const shortfallResult = await db.query(
    `SELECT COUNT(*) as cnt
     FROM (
       SELECT part_id, COALESCE(SUM(qty), 0) as wh_qty
       FROM stock WHERE location_type = 'warehouse'
       GROUP BY part_id
     ) sq
     JOIN parts p ON p.id = sq.part_id
     WHERE p.is_active = 1 AND sq.wh_qty < p.min_stock_level`,
  );
  const shortfallCount = shortfallResult.values[0]?.cnt ?? 0;

  // Pending tasks (staged items count)
  const pendingResult = await db.query(
    `SELECT COUNT(*) as cnt FROM stock WHERE location_type = 'pulled' AND qty > 0`,
  );
  const pendingCount = pendingResult.values[0]?.cnt ?? 0;

  return {
    stock_health_pct: healthPct,
    total_units: totalUnits,
    shortfall_count: shortfallCount,
    pending_task_count: pendingCount,
  };
}

/** Get recent stock activity */
export async function getRecentActivity(limit = 10): Promise<ActivitySummary[]> {
  const db = await getDb();
  const result = await db.query(
    `SELECT sm.id, sm.movement_type, sm.qty, sm.created_at,
       p.part_number, p.description as part_desc,
       u.display_name as performer_name,
       sm.from_location_type, sm.to_location_type
     FROM stock_movements sm
     JOIN parts p ON p.id = sm.part_id
     LEFT JOIN users u ON u.id = sm.performed_by
     ORDER BY sm.created_at DESC
     LIMIT ?`,
    [limit],
  );

  return result.values.map((row) => ({
    id: row.id as number,
    summary: `${row.performer_name ?? 'Unknown'} ${row.movement_type}d ${row.qty}× ${row.part_number} (${row.from_location_type ?? 'external'} → ${row.to_location_type ?? 'external'})`,
    movement_type: row.movement_type as string,
    performer_name: row.performer_name as string | null,
    created_at: row.created_at as string,
  }));
}

/** Get inventory grid with search and filters */
export async function getInventoryGrid(opts?: {
  search?: string;
  category_id?: number;
  status?: string;
  sort?: string;
  limit?: number;
  offset?: number;
}): Promise<{ items: InventoryItem[]; total: number }> {
  const db = await getDb();
  const conditions: string[] = ['p.is_active = 1'];
  const params: any[] = [];

  if (opts?.search) {
    conditions.push('(p.part_number LIKE ? OR p.description LIKE ?)');
    const term = `%${opts.search}%`;
    params.push(term, term);
  }

  if (opts?.category_id) {
    conditions.push('p.category_id = ?');
    params.push(opts.category_id);
  }

  const where = `WHERE ${conditions.join(' AND ')}`;
  const sortMap: Record<string, string> = {
    name: 'p.part_number',
    qty: 'warehouse_qty',
    status: 'stock_status',
  };
  const sort = sortMap[opts?.sort ?? 'name'] ?? 'p.part_number';
  const limit = opts?.limit ?? 100;
  const offset = opts?.offset ?? 0;

  const countResult = await db.query(
    `SELECT COUNT(*) as cnt FROM parts p ${where}`,
    params,
  );

  const result = await db.query(
    `SELECT p.id, p.part_number, p.description,
       c.name as category_name, b.name as brand_name,
       COALESCE(sq.wh_qty, 0) as warehouse_qty,
       p.min_stock_level as min_stock,
       p.max_stock_level as max_stock,
       p.reorder_point,
       CASE
         WHEN COALESCE(sq.wh_qty, 0) < p.min_stock_level THEN 'below_min'
         WHEN COALESCE(sq.wh_qty, 0) > p.max_stock_level THEN 'above_max'
         ELSE 'healthy'
       END as status,
       sq.last_moved
     FROM parts p
     LEFT JOIN categories c ON c.id = p.category_id
     LEFT JOIN brands b ON b.id = p.brand_id
     LEFT JOIN (
       SELECT part_id,
         SUM(qty) as wh_qty,
         MAX(updated_at) as last_moved
       FROM stock WHERE location_type = 'warehouse'
       GROUP BY part_id
     ) sq ON sq.part_id = p.id
     ${where}
     ORDER BY ${sort} ASC
     LIMIT ? OFFSET ?`,
    [...params, limit, offset],
  );

  // Apply status filter post-query if needed
  let items = result.values as InventoryItem[];
  if (opts?.status) {
    items = items.filter((i) => i.status === opts.status);
  }

  return {
    items,
    total: countResult.values[0]?.cnt ?? 0,
  };
}

/** Get detailed stock info for a specific part across all locations */
export async function getPartStockDetail(partId: number): Promise<PartStockDetail | null> {
  const db = await getDb();

  const partResult = await db.query(
    'SELECT id, part_number, description FROM parts WHERE id = ?',
    [partId],
  );
  const part = partResult.values[0];
  if (!part) return null;

  const stockResult = await db.query(
    `SELECT s.location_type, s.location_id, s.qty,
       sup.name as supplier_name,
       CASE s.location_type
         WHEN 'warehouse' THEN 'Warehouse'
         WHEN 'pulled' THEN 'Pulled/Staged'
         WHEN 'truck' THEN (SELECT name FROM vehicles WHERE id = s.location_id)
         WHEN 'job' THEN (SELECT job_name FROM jobs WHERE id = s.location_id)
         ELSE s.location_type
       END as location_name
     FROM stock s
     LEFT JOIN suppliers sup ON sup.id = s.supplier_id
     WHERE s.part_id = ? AND s.qty > 0
     ORDER BY s.location_type ASC, s.location_id ASC`,
    [partId],
  );

  const locations = stockResult.values as StockLocation[];
  const totalQty = locations.reduce((sum, loc) => sum + loc.qty, 0);

  return {
    part_id: part.id as number,
    part_number: part.part_number as string,
    description: part.description as string,
    locations,
    total_qty: totalQty,
  };
}

/** Search parts by number or description */
export async function searchParts(
  query: string,
  limit = 20,
): Promise<{ id: number; part_number: string; description: string; warehouse_qty: number }[]> {
  const db = await getDb();
  const term = `%${query}%`;
  const result = await db.query(
    `SELECT p.id, p.part_number, p.description,
       COALESCE((SELECT SUM(qty) FROM stock WHERE part_id = p.id AND location_type = 'warehouse'), 0) as warehouse_qty
     FROM parts p
     WHERE p.is_active = 1 AND (p.part_number LIKE ? OR p.description LIKE ?)
     ORDER BY p.part_number ASC
     LIMIT ?`,
    [term, term, limit],
  );
  return result.values as any[];
}
